### ISSUE TITLE
Routing extension requirement of routes having to exist can conflict with fuzzy route matching

### ISSUE THREAD

**Issue by ndm2** [icings/menu#2]
The routing extension converts URL arrays to URL strings, which requires the routes to exist, otherwise a `MissingRouteException` is being thrown. This can conflict with fuzzy route matching, in case one wants to match on a part of a URL for wich no connected route exists.

If for example the following routes are connected:

```php
$routes->connect('/members/about', ['controller' => 'Members', 'action' => 'about']);
$routes->connect('/members/view', ['controller' => 'Members', 'action' => 'view']);
```

and one wants to fuzzy match on

    ['controller' => 'Members']

so that all member URLs are being matched, this wont work, as no matching route for that URL array is connected.

Until this is fixed, one workaround would be to create dummy routes, ideally ones that cannot be reached, for example because they are being catched by other, previously connected routes, like:

```php
$routes->connect('/members/about', ['controller' => 'Members', 'action' => 'about']);
$routes->connect('/members/view', ['controller' => 'Members', 'action' => 'view']);
// ...

// this will never be reached, as the previous /members/about route will
// catch the request, but it satsfies the routing extension requirements
$routes->connect('/members/about', ['controller' => 'Members']);
```

**Comment by dereuromark** [comment]
I would recommend adding a URL routes matching, that would solve quite a few of these issues without workaround, right?
You have definitely more options on those arrays than on string comparison.

**Comment by ndm2** [comment]
I don't really want to drop string URL matching support. The decision to move string URL generation into the routing extension was to avoid lots of unneccesary `Router::url()` calls in the voters themselves, but later this turned out to be largely a caching problem, and unfortunately I forgot to move string URL generation once the caching problem was fixed.

I was thinking about alternative approaches, but I'll probably go for moving URL generation into the URL voter (which unfortunately isn't backwards compatible), though this would currently cause one unneccesary `Router::url()` call per item when using the URL voter, ideally I'll find a way to avoid that.

Regardless of that, adding a non-fuzzy URL array based voter might be an option. Not sure though what advantages (apart from probably a small performance gain) comparing URL arrays instead of string URLs would have? I'll have to think about that.

**Comment by ndm2** [comment]
For those who'd like to test the new behavior, check out the `2.0` branch. URL array conversion (for the `routes` option) has been moved into the URL voter.

**Comment by ndm2** [comment]
This is fixed and will be available in the soon to be released 2.0
